mirgecom/transport.py

```diff
@@ -0,0 +1,212 @@
+($\mu_{B}$), the thermal conductivity ($\kappa$), and the species diffusivities
+($d_{\alpha}$).
+
+Two models are currently implemented, the :class:`SimpleTransport` model is
+for uniform and constant transport properties and the :class:`PowerLawTransport`
+is for transport properties that are power-law temperature-dependent.
+
+.. autoclass:: TransportDependentVars
+.. autoclass:: TransportModel
+.. autoclass:: SimpleTransport
+.. autoclass:: PowerLawTransport
+"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from dataclasses import dataclass
+import numpy as np
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from mirgecom.fluid import ConservedVars
+from mirgecom.eos import GasEOS
+
+
+@dataclass
+class TransportDependentVars:
+    """State-dependent quantities for :class:`TransportModel`.
+
+    Prefer individual methods for model use, use this
+    structure for visualization or probing.
+
+    .. attribute:: bulk_viscosity
+    .. attribute:: viscosity
+    .. attribute:: thermal_conductivity
+    .. attribute:: species_diffusivity
+    """
+
+    bulk_viscosity: np.ndarray
+    viscosity: np.ndarray
+    thermal_conductivity: np.ndarray
+    species_diffusivity: np.ndarray
+
+
+class TransportModel:
+    r"""Abstract interface to thermo-diffusive transport model class.
+
+    Transport model classes are responsible for
+    computing relations between fluid or gas state variables and
+    thermo-diffusive transport properties for those fluids.
+
+    .. automethod:: bulk_viscosity
+    .. automethod:: viscosity
+    .. automethod:: thermal_conductivity
+    .. automethod:: species_diffusivity
+    .. automethod:: viscosity2
+    """
+
+    def bulk_viscosity(self, eos: GasEOS, cv: ConservedVars):
+        r"""Get the bulk viscosity for the gas (${\mu}_{B}$)."""
+        raise NotImplementedError()
```

Comment:
Hmm... EOS and transport object interdependency strikes me as kind of an architectural red flag. Is there a way they could be set up to avoid this?

I don't have a good sense of what they need from each other. Can you explain what (in the most general case) the EOS needs from the transport model and vice versa?

Reply:
From the physics point of view, they belong together.  The EOS is the part of the gas model that applies to gases in equilibrium, and the transport properties are part of the gas model that applies to non-equilibrium states.  Arguably the transport properties are more fundamental, but in our macroscopic treatment of the gas - they are both just property models.  For our simulations, inviscid flow simulations require only EOS, and viscous ones require EOS+transport.

From a data dependency point of view:  Our transport model considers the transport properties to be a function of the macroscopic thermodynamic state, so the transport model needs the EOS. But the EOS-proper does not need the transport model parameters at all. 

From a code architectural point of view, things are a little more gray.  I like having them together because I pass one argument that represents the physical model of the gas.  Arguably, it should be called something other than just EOS, like GasModel or some such that includes the EOS and the transport model.  But making it "part of" the EOS object is easily understandable by the users, and imo is a minor foible.  

On the other hand, one certainly _could_ consider the transport properties as a separate construct and pass it around as a separate thing. The data dependency does not prevent this type of a design. IMO, such a design has a more complicated interface and does not promote the understanding that the physical gas model parameters are encapsulated by the EOS/Transport pair. 

One other thing that I hope will help you consider them as one:  Pyrometheus is exactly the sort of package that could easily just be called `GasModel` or some such, and it underlies both the EOS and the transport properties.   It does not yet support transport properties - but support for that is coming and when it is ready - we will readily switch our transport models to a pyrometheus-based one, and currently the EOS is our interface to Pyrometheus and both the EOS and transport need to talk to the *same* Pyrometheus object - so it makes sense that we house them under the same roof.

Reply:
> From the physics point of view, they belong together. The EOS is the part of the gas model that applies to gases in equilibrium, and the transport properties are part of the gas model that applies to non-equilibrium states. Arguably the transport properties are more fundamental, but in our macroscopic treatment of the gas - they are both just property models. For our simulations, inviscid flow simulations require only EOS, and viscous ones require EOS+transport.
> 
> From a data dependency point of view: Our transport model considers the transport properties to be a function of the macroscopic thermodynamic state, so the transport model needs the EOS. But the EOS-proper does not need the transport model parameters at all.

Got it. Thanks!

> From a code architectural point of view, things are a little more gray. I like having them together because I pass one argument that represents the physical model of the gas. Arguably, it should be called something other than just EOS, like GasModel or some such that includes the EOS and the transport model. But making it "part of" the EOS object is easily understandable by the users, and imo is a minor foible.

Actually, I like this idea. 🙂 Could we make something like a `GasModel` class that takes in a `GasEOS` and (optionally) a `TransportModel` and provides an interface to the fluid operators containing the quantities from both? That would remove the circular EOS/Transport dependency, and it would make me feel better about the code structure. I don't think we would need to create a whole hierarchy of `GasModel`s; seems like we could get away with it being a single wrapper class (unless we wanted to separate by inviscid/viscous maybe?).

Reply:
> Actually, I like this idea. 🙂 Could we make something like a `GasModel` class that takes in a `GasEOS` and (optionally) a `TransportModel` and provides an interface to the fluid operators containing the quantities from both? That would remove the circular EOS/Transport dependency, and it would make me feel better about the code structure. I don't think we would need to create a whole hierarchy of `GasModel`s; seems like we could get away with it being a single wrapper class (unless we wanted to separate by inviscid/viscous maybe?).

Yes, I think we should do this. It would be my preference to tackle this in a dedicated effort, however.  I believe it is the right code structure, and would allow us to also separate out chemistry, which is also currently wedged into EOS.  

Having this stuff in EOS is not technically incorrect, so imo there is would be no harm in pursuing the code structure / architecture refactor as a separate effort.

Reply:
👍 Sounds good.